### PR TITLE
Kevin jupyterhub

### PR DESCRIPTION
**Pull request by KevinFHartmann**
Here's my first draft... I don't really like the way it flows yet, and it doesn't say everything that I want, but it's a start...

### CODE REVIEW

File: _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    version: small
+    title: Software Engineering Blog
+    image: header-logo-crop.png
+    icon: icon-blog
+---
+<image src="{{ site.urlimg }}/kevin_hartmann/jupiter.jpg" alt="Weather vortex pattterns on planet Jupiter" />
+
+Hey, wait! That's the surface of **Jupiter**, the planet! We're talking about **Jupyter**, with a 'y', as in **Jupyter Notebooks** and their multi-user counterpart, **Jupyterhub**! 
+
+<image src="{{ site.urlimg }}/kevin_hartmann/jupyter.png" alt="Jupyter project logo" /> Yeah, those guys! That's much better!
```
Comment (Fgerthoffert) [inline _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md]: Could you change the image to one with less artifacts (or reduce its size) ?

File: _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+This blog post is about how I set up our Jupyterhub service at jupyterhub.cancercollaboratory.org. It's aimed at a technical audience, so if you don't have that kind of background, don't be afraid to skip over parts that you don't understand right away, and come back to them later. That's how I learned most of what I know about computers; as far as I know, it's how most people still do. With that said, let's carry on into our exploration into the wild and tumultuous world of Jupyterhub!
+
+Our Jupyterhub service is hosted on a virtual machine inside an Openstack cluster, and we naturally wanted to be able to re-build the whole thing from scratch without having to work too hard at it. We use *ansible* for automation of these sorts of things, so I created an ansible playbook to set up the virtual machine, and set up jupyterhub to run as a docker container within the virtual machine.
```
Comment (Fgerthoffert) [inline _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md]: Can you add a section on background regarding the hosting part, why one VM, and so on. Nothing large, just a short paragraph.

File: _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+2. Ensure that the virtual machine is running python (so ansible can run)
+3. Create the users and permissions on the virtual machine 
+**and**
+4. Set up the jupyterhub server
+
+In order to accomplish these four tasks, I created the following five ansible roles:
+
+1. **openstack:** Accomplishes the OpenStack setup task
+2. **users**: Creates the users and groups necessary to run docker, and the jupyterhub service itself.
+3. **certbot**: Runs "certbot" to create a signed SSL certificate for the site <link>certbot</link>, and sets up a cron job that will automatically re-generate the certificate before it expires.
```
Comment (Fgerthoffert) [inline _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md]: The link tag might not be recognized properly, it shows up in the rendered file

File: _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+4. **docker**: Installs docker on the virtual machine, and sets up our dockerized jupyterhub system run from a single command.
+5. **jupyterhub**: Adds a service configuration for jupyterhub that launches our dockerized jupyterhub system when the VM boots up, which lets start and stop it like any other service.
+
+### If at first you don't succeed... 
+
+Each step in our ansible playbook is designed to be re-run if the a play fails to complete for any reason. There can be a lot of different steps to configure, but we keep things organized by putting all of that configuration information into two files that do only one thing: set ansible variables.
```
Comment (Fgerthoffert) [inline _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md]: extra "a" in "if the a play fails to complete"

File: _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+For us, using docker simplifies development and testing. By using Docker, I can separate many of the issues related to running the virtual machine (and it's associated networking, services, security concerns, and so on) from the configuration of the Jupyterhub system itself.
+
+And by using Docker, I and other developers can spin up a copy of Jupyterhub on our computer, make any changes we want to the look and feel or operating features that our Jupyterhub system should have, and then deploy it.
+
+Within the Jupyterhub system, I further isolated the section of the python module that sets up all the web related settings for Jupyterhub (the css modules, the template files, and so on), by providing a symbolic link from the service directory to the relavent section of the python library module. That way, we can isolate the changes we make to the Jupyterhub front end from the python code that launches those changes; making it easy to change the web side of the system without worrying too much about the code itself.
+
+# What's this button do? 
+<image src="{{ site.urlimg }}/kevin_hartmann/cat.gif" />
```
Comment (Fgerthoffert) [inline _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md]: The cat gif is great but it tickles the eye a bit, can you get either a static image or one either smaller or less extreme in color transitions?

File: _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+  **and** 
+- *reset.sh* : A convenience script for development that brings down our service; deletes our old jupyterhub image; builds the new image, and brings the service back up
+
+So far, so good. But what are the contents of the *hub* and *notebook* directories? Let's look at our requirements for our Jupyterhub service first, so that we know what files we'll need for the *hub* directory, and then we'll look at the *notebook* directory.
+
+### Jupyterhub, at your service... 
+
+Our jupyterhub service's job is to present the main login screen, authenticate our end users using single sign on, and to spin up the right docker container with the right docker storage volume mounted for our end-user's account. 
+
+To make matters slightly more complicated, we have our own custom docker authenticator (which verifies that our end users have DACO access) that we also want to run, and we have some other look and feel settings that we would like to customize for the sign-on portion of our end user's web experience.
```
Comment (Fgerthoffert) [inline _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md]: Can you add a link to DACO ?

File: _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+<image src="{{ site.urlimg }}/kevin_hartmann/calculator.jpg" /> 
+
+We use Docker to run Jupyterhub so that we can abstract away specific details bout about how Jupyerhub is being run, like where the SSL keys are located, or which url to use for the authentication service. 
+
+That way, we can deploy Jupyterhub anywhere, just by tweaking the environment we run it in, and specifically, we can run it locally in development mode before deploying it onto our production servers. 
+
+We use ansible to take care of the details of setting up a virtual machine, setting up users and permissions, setting up environment variables with passwords and access tokens that docker will use to run, and so on... 
+
+We use multiple docker images, and a docker volume, copies the data from the docker image into the volume, run docker-compose to run jupyterhub, which launches jupyter notebooks with the volume for the conda distribution mount on /opt, which lets our users enjoy the fun of Jupyter notebooks without all of the hassle of setting up Jupyterhub. 
+
```
Comment (Fgerthoffert) [inline _posts/kevin_hartmann/2018-03-19-Drops-of-Jupyter.md]: In your conclusion/summary, can you add links to the actual source or more resources people could use to get started right away ?

**Review by Fgerthoffert — COMMENTED** [review]
Great blog post

**Review by andricDu — APPROVED** [review]

**Comment by andricDu** [comment]
@KevinFHartmann do you mind just updating the date of the post to today and we will release.

**Comment by KevinFHartmann** [comment]
I thought we said we were holding off until next week, because we wanted Grant's post to go first?

________________________________

**Pull request merged**